### PR TITLE
[Gradient Compression] Add a documentation page for DDP communication hooks

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51715 [Gradient Compression] Add a documentation page for DDP communication hooks**

Add a documentation page for DDP communication hooks.

image

Facebook:
Full HTML file: P172254228

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

### CODE REVIEW

File: docs/source/ddp_comm_hooks.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
+DDP Communication Hooks
+=======================
+
+DDP communication hook is a generic interface to control how to communicate
```
Comment (rohan-varma) [inline docs/source/ddp_comm_hooks.rst:4]: Might be good to explicitly note that these hooks are used in conjunction with DistributedDataParallel in the first paragraph and note that it will override the allreduce.

File: docs/source/ddp_comm_hooks.rst [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+gradients across workers.
+A few built-in communication hooks are provided,
+and users can easily apply any of these hooks to optimize communication.
+Besides, the hook interface can also support user-defined communication
+strategies for more advanced use cases.
+
+.. warning ::
+    DDP communication hooks can only support single process single device mode
```
Comment (rohan-varma) [inline docs/source/ddp_comm_hooks.rst:12]: As a side note, may be good to double check that we throw the appropriate errors if wrong backend or mode is used.

File: docs/source/ddp_comm_hooks.rst [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    on NCCL backend.
+
+How to Use A Communication Hook?
+--------------------------------
+
+To use a communication hook, the user just needs to let the DDP model register
+the hook before the training loop.
+
+.. automethod:: torch.nn.parallel.DistributedDataParallel.register_comm_hook
```
Comment (rohan-varma) [inline docs/source/ddp_comm_hooks.rst:21]: Can you also add a code example for how a user might define an allreduce hook, and then call it with a state object?

File: docs/source/ddp_comm_hooks.rst [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+The performance is `on par with <https://observablehq.com/@tvogels/powersgd-benchmark>`_
+the implementation in the original `paper <https://arxiv.org/abs/1905.13727>`_.
+
+PowerSGD State
+^^^^^^^^^^^^^^^^
+
+.. currentmodule:: torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook
+.. autoclass:: PowerSGDState
+
+PowerSGD Hooks
```
Comment (rohan-varma) [inline docs/source/ddp_comm_hooks.rst:46]: Maybe Add a note that there are batched and unbatched versions, as well as the tradeoffs between them?

**Review by rohan-varma — COMMENTED** [review]
Looks good, but can you link/screenshot a page of how the documentation renders currently? Thanks for adding this!

File: docs/source/ddp_comm_hooks.rst [hunk 5]
```diff
@@ -0,0 +1,64 @@
+DDP Communication Hooks
+=======================
```
Comment (pritamdamania87) [inline docs/source/ddp_comm_hooks.rst:2]: We should add a warning at the top of the page mentioning this feature is experimental. Similar to what we have in https://pytorch.org/docs/master/pipeline.html

File: docs/source/ddp_comm_hooks.rst (hunk 3 above)
Comment (pritamdamania87) [reply docs/source/ddp_comm_hooks.rst:21]: I believe the docstrings for `register_comm_hook` do have a couple of examples.

File: docs/source/ddp_comm_hooks.rst [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+.. automethod:: torch.nn.parallel.DistributedDataParallel.register_comm_hook
+
+Default Communication Hooks
+---------------------------
+
+Default communication hooks are simple **stateless** hooks, so the input state
+in ``register_comm_hook`` is either a process group or ``None``.
+
+.. automodule:: torch.distributed.algorithms.ddp_comm_hooks.default_hooks
+    :members:
```
Comment (pritamdamania87) [inline docs/source/ddp_comm_hooks.rst:30]: Just wanted to double check this doesn't pull in the private methods from `default_hooks` right? Also, we probably should not pull in `allreduce_fut` which is a public method (should it be private)?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 7]
```diff
@@ -162,38 +164,48 @@ def maybe_increase_iter(self, bucket):
...
+        3.7. Computes each M among all the high-rank tensors, which is approximately equal to PQ^T.
+
+    Note that this communication hook enforces vanilla allreduce for the first ``state.start_powerSGD_iter`` iterations.
     This can not only allow the user to have a finer tuning over the tradeoff between speedup and accuracy,
     but also help abstract away some complexity of the internal optimization of DDP for future communication hook developers.
 
-    TODO(wayi@): The above procedure does two matmul+allreduce steps per iteration --
+    TODO: The above procedure does two matmul+allreduce steps per iteration --
     one left multiplication and one right multiplication.
     For warm-start, can take one such step at a time, and alternate between them.
```
Comment (pritamdamania87) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:204]: TODOs should not be in the docs, this should probably be a comment in the code that is not visible to the user in docs.

File: torch/nn/parallel/distributed.py [hunk 8]
```diff
@@ -1063,9 +1064,6 @@ def register_comm_hook(self, state: object, hook: callable):
             ``register_comm_hook`` API, but it is subject to some subtle differences compared
             to ``torch.futures.Future``.
 
-        .. warning ::
-            DDP communication hook is experimental and subject to change.
-
```
Comment (pritamdamania87) [inline torch/nn/parallel/distributed.py:1068]: Why was this warning removed? We apply the "experimental and subject to change" warning to all "beta" features and I believe ddp comm hook is still "beta" right?

Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:1068]: I see. Recovered.

File: docs/source/ddp_comm_hooks.rst (hunk 6 above)
Comment (SciPioneer) [reply docs/source/ddp_comm_hooks.rst:30]: `_ allreduce_fut` is a private method, and only `allreduce_hook` and `fp16_compress_hook` are exposed.

File: docs/source/ddp_comm_hooks.rst (hunk 3 above)
Comment (SciPioneer) [reply docs/source/ddp_comm_hooks.rst:21]: Yes, already `allreduce_hook` already has an example in the docstrings.

File: docs/source/ddp_comm_hooks.rst (hunk 4 above)
Comment (SciPioneer) [reply docs/source/ddp_comm_hooks.rst:46]: Already covered by the docstrings. Copied it here:

This variant is faster than powerSGD_hook() that runs layer-wise gradient compression, but it usually results in a much lower accuracy, unless matrix_approximation_rank in the state is 1. Increasing matrix_approximation_rank may not necessarily increase the accuracy, because batching per-parameter tensors without column/row alignment can destroy low-rank structure. Therefore, the user should always consider powerSGD_hook() first, and only consider this variant when a satisfying accuracy can be achieved when matrix_approximation_rank is 1.

File: docs/source/ddp_comm_hooks.rst (hunk 2 above)
Comment (SciPioneer) [reply docs/source/ddp_comm_hooks.rst:12]: 1) single-process single-device mode is already checked in `reducer.cpp`.
2) will add another check for the backend in a separate PR.

File: docs/source/ddp_comm_hooks.rst (hunk 1 above)
Comment (SciPioneer) [reply docs/source/ddp_comm_hooks.rst:4]: Updated the first sentence as:

DDP communication hook is a generic interface to control how to communicate gradients across workers and override the vanilla allreduce in `DistributedDataParallel`.

File: docs/source/ddp_comm_hooks.rst [hunk 9]
```diff
@@ -2,12 +2,15 @@ DDP Communication Hooks
 =======================
 
 DDP communication hook is a generic interface to control how to communicate
-gradients across workers.
+gradients across workers by overriding the vanilla allreduce in ``DistributedDataParallel``.
```
Comment (pritamdamania87) [inline docs/source/ddp_comm_hooks.rst:5]: nit: Might be nice to have a clickable link to DDP here: https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.

**Comment by SciPioneer** [comment]
> Looks good, but can you link/screenshot a page of how the documentation renders currently? Thanks for adding this!

**Review by pritamdamania87 — APPROVED** [review]

**Comment by ngimel** [comment]
Reverting, it broke the doc build.
